### ISSUE TITLE
App freezes/crashes when trying to export certain keys

### ISSUE THREAD

**Issue by donpatzone** [chris2511/xca#593]
Hi,

I am currently running into an issue where the app locks up when I try to export certain private keys. I ran into this issue with v2.6.0 on Windows 11. The key in question has its source stated as "Legacy Database". I tried exporting another one and this one worked fine. I already tried upgrading to version 2.7.0 but the issue still persists. Same with using a different Windows machine as well.

Are there any debugging capabilities for the Windows platform in order to provide more information?

Best Regards

**Comment by chris2511** [comment]
you can start xca on the commandline with the --verbose option and your console will be full of more or less helpful messages.
If the app "locks up" it means it becomes unresponsible and must be killed via task-manager ?

**Comment by donpatzone** [comment]
correct. Need to kill it in the task manager, as it is unresponsible. Also, in the meantime I was able to test the database on a Mac. Same behavior. The app locks up and I need to kill it.

With the verbose flag I got the following as soon as I try any sort of key retrieval (copy to clipboard/export key)
```
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11700 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11700 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11701 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11701 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11701 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
 11701 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 11701 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[3]) - Rows selected: -1"
```

And this fires constantly several times per second.

**Comment by chris2511** [comment]
`myencKey.size() 0` Indicates that the private key is gone ???
The app-locking during export is a bug in my code i found now - triggered by empty private key data.
But it won't solve the problem that those private keys are "gone"

Please use the sqlitebrowser or sqlite3 cmdline app, open the database and look at the *private_keys* table:
`select * from private_keys;`
It should contain several entries and at least one with

3|0|&lt;some1000 bytes base64&gt;

**Comment by donpatzone** [comment]
I can confirm this to be the case. The keys show no value for the problematic entries (index 3 & 12). 
![db_priv_keys](https://github.com/user-attachments/assets/97d5563b-2f91-4a4c-a8a9-bef983c26c3b)

The app still says it is available.
![grafik](https://github.com/user-attachments/assets/ea90116c-92ab-4764-8baa-2023712446a3)

That is really strange. Is there a way to recover them? Because I don't understand how they can get lost.

**Comment by chris2511** [comment]
XCA says it is available, because it has an entry in the `private_keys` table. Only if it tries to accees it (for export or signing), it fails.
If you do not have a copy of the database, there is no way to recover them, sorry.

I take this as a serious issue. This is an "illegal" situation.
If the private key is removed by XCA, then the whole line is gone.

Do you have any hints what special things you might have done with those keys ?
Are they imported or generated?
Have you ever successfully exported them once in the past?
Is your database on a network share ?

**Comment by donpatzone** [comment]
* Nothing special that I know of. The database is already being used for several years. Those are one of the older entries (dating back at least to 2018). They are flagged as Legacy database. I also see a file next to it with {shrink} in its name. 
* Yes they were exported successfully. I now had to use them again as a certificate renewal is due and that is where I noticed the issue.
* The file itself is not on a share, but it is synced to a nextcloud instance. File access always happens locally and only on this machine, so I am confident that there was no other operation done to the file from somewhere else.

In the meantime I found a backup from before the change to SQLite. I now have the keys. I deleted the faulty entries and re-imported them. Everything is working again now.

**Comment by donpatzone** [comment]
I was a bit premature with my assumption. Has to do with the lack of properly going through the database and seeing 2 empty entries I jumped to a conclusion way too early. 

I noticed that I was unable to create a new request for a completely different entry, the app froze again. Doing a debug the same error is fired. 

```
 13706 Debug: .\lib\sql.cpp:208 XSqlQuery::exec "QUERY: .\\lib\\pki_evp.cpp:750 (SELECT private FROM private_keys WHERE item=?[12]) - Rows selected: -1"
 13706 Debug: .\lib\pki_evp.cpp:505 pki_evp::tryDecryptKey myencKey.size() 0
 ```
However when I directly tried to export the key it was working.

The error gave a clue that now index 12 is the issue.

I opened the database again with the sqlite browser and noticed indeed row 12 is still an empty line in the `private_keys` table. I noticed there is another table called `view_private`.
This one gave me the name of the entry instead of a base64 string. There I saw that actually one of the 2 keys I thought was corrupt is indeed there and that a completely different entry is empty. This one was luckily an unused one, which btw was also a legacy entry. I went ahead and deleted this unused key. Afterwards I confirmed that the database did not contain any empty lines.

Now the create request function also works again.

I am assuming that this is due to a positional issue. Meaning operations with keys that are above row 12 would work and everything below would not - that is an assumption, I did not confirm with a test.

**Comment by chris2511** [comment]
The `view_private` actually is not a table, but a view. It combines the `items.name` and `private_keys.private` where `items.id = private_keys.item`.

This allows to easily select private keys by internal name.

I will close this issue with a fix, where I handle empty keys more sane with an error message instead of freezing in an endless loop.
I will also add a double-check to have an encrypted private key with size > 0 when writing it to the database.

I'm however still worried how this could happen in the first place.
Right now i again tested updating old databases with legacy keys and all worked fine.

So adding a safety-net is all I can do for now.

**Comment by donpatzone** [comment]
Thanks, also fyi. Maybe this is only on my side, but I just had a weird behavior. You see I have a root CA, which as 2 intermediate CAs that share the same CN, due to the one still using an older signature algorithm. I just saw that some device certificates entries there were listed under the older intermediate when I checked the chain of trust. Deleting and reimporting them fixed it and now they are under the correct (newer) intermediate. Currently I assume this has something to do with my issues.
